cmd/algokey/part.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/algorand/go-algorand/util/db"
+)
+
+var partKeyfile string
+var partFirstRound uint64
+var partLastRound uint64
+var partKeyDilution uint64
+var partParent string
+
+var partCmd = &cobra.Command{
+	Use:   "part",
+	Short: "Manage participation keys",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, args []string) {
+		// If no arguments passed, we should fallback to help
+		cmd.HelpFunc()(cmd, args)
+	},
+}
+
+var partGenerateCmd = &cobra.Command{
+	Use:   "generate",
+	Short: "Generate participation key",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, _ []string) {
+		if partLastRound < partFirstRound {
+			fmt.Fprintf(os.Stderr, "Last round %d < first round %d\n", partLastRound, partFirstRound)
+			os.Exit(1)
+		}
+
+		if partKeyDilution == 0 {
+			partKeyDilution = 1 + uint64(math.Sqrt(float64(partLastRound-partFirstRound)))
+		}
+
+		var err error
+		var parent basics.Address
+		if partParent != "" {
+			parent, err = basics.UnmarshalChecksumAddress(partParent)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Cannot parse parent address %s: %v\n", partParent, err)
+				os.Exit(1)
+			}
+		}
+
+		partdb, err := db.MakeErasableAccessor(partKeyfile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot open partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		partkey, err := account.FillDBWithParticipationKeys(partdb, parent, basics.Round(partFirstRound), basics.Round(partLastRound), partKeyDilution)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot generate partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		printPartkey(partkey)
+	},
+}
+
+var partInfoCmd = &cobra.Command{
+	Use:   "info",
+	Short: "Print participation key information",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, _ []string) {
+		partdb, err := db.MakeErasableAccessor(partKeyfile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot open partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		partkey, err := account.RestoreParticipation(partdb)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot load partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		printPartkey(partkey)
+	},
+}
+
+var partReparentCmd = &cobra.Command{
+	Use:   "reparent",
+	Short: "Change parent address of participation key",
+	Args:  cobra.NoArgs,
+	Run: func(cmd *cobra.Command, _ []string) {
+		parent, err := basics.UnmarshalChecksumAddress(partParent)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot parse parent address %s: %v\n", partParent, err)
+			os.Exit(1)
+		}
+
+		partdb, err := db.MakeErasableAccessor(partKeyfile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot open partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		partkey, err := account.RestoreParticipation(partdb)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot load partkey database %s: %v\n", partKeyfile, err)
+			os.Exit(1)
+		}
+
+		partkey.Parent = parent
```

Comment:
Rather than breaking the assignment and persistence, can we change PersistNewParent() to ReplaceParent(parent)?

Reply:
I was going for that originally, but for historical reasons, all current methods on `account.Participation` are by value rather than by pointer.  So, `ReplaceParent()` would not have any apparent effect on the caller's partkey object (in particular, the subsequent `printPartkey()` would print the old parent), even though it could write the new parent to the partkey db..  Modifying the parent in-place seemed cleaner than having a mix of by-value and by-pointer methods on `account.Participation`, and changing all of them to by-pointer seemed unnecessarily involved.